cpp/src/gandiva/precompiled/string_ops.cc

```diff
@@ -121,10 +122,16 @@ int32 utf8_length(int64 context, const char* data, int32 data_len) {
   int count = 0;
   for (int i = 0; i < data_len; i += char_len) {
     char_len = utf8_char_length(data[i]);
-    if (char_len == 0) {
+    if (char_len == 0 || i + char_len > data_len) {  // invalid byte or incomplete glyph
       set_error_for_invalid_utf(context, data[i]);
       return 0;
     }
+    for (int j = 1; j < char_len; ++j) {
+      if (data[i + j] > 0) {  // bytes following head-byte of glyph
```

Comment:
This is a great catch, thanks! Could you add unit tests (for `utf8_length`) for these error cases?
```suggestion
      if ((data[i + j] & 0xE0) != 0xC0) {  // bytes following head-byte of glyph
```

Reply:
shouldn't it be data[I+j] & 0xC0 != 0x80? we want to check if the byte starts with 10xx

Reply:
Yes you're right. I've slipped a bit, so to speak